lib/index.ts

```diff
@@ -0,0 +1,561 @@
+		reply: (params: MessageRequest.Reply, callback?: Callback<Status>): Promise<Status> =>
+			this.httpCall({ path: 'conversations/<conversation_id>/messages', reqType: 'POST' }, params, callback),
+		send: (params: MessageRequest.Send, callback?: Callback<ConversationReference>):
+			Promise<ConversationReference> => this.httpCall({ path: 'channels/<channel_id>/messages',
+			reqType: 'POST' }, params, callback),
+	};
+
+	private apiKey: string;
+
+	constructor(apiKey: string) {
+		// Key.
+		this.apiKey = apiKey;
+	}
+
+	private httpCall(details: Request, params: any, callback?: InternalCallback): Promise<any | void>  {
+		const requestOpts = {
+			body: (params) ? params : {},
+			headers: {
+				Authorization: `Bearer ${this.apiKey}`
+			},
+			json: true,
+			method: details.reqType,
+			url: `${URL}/${this.formatPath(details.path, params)}`
+		};
+
+		// Make the request.
+		return request(requestOpts).promise().catch((error: any) => {
+			// Format this into something useful, if we can.
+			throw new FrontError(error);
+		}).asCallback(callback);
+	}
+
+	private formatPath(path: string, data: RequestData = {}): string {
+		let newPath = path;
+		const reSearch = (re: RegExp, operation: (matches: RegExpMatchArray) => void) => {
+			let matches = path.match(re);
+			if (matches) {
+				operation(matches);
+			}
+		};
+
+		// Find the mandatories. If we don't get them, then we error.
+		reSearch(/<(.*?)>/g, (mandatoryTags: RegExpMatchArray) => {
+			_.map(mandatoryTags, (tag) => {
+				const tagName = tag.substring(1, tag.length -1);
+				if (!data[tagName]) {
+					throw new Error(`Tag ${tag} not found in parameter data`);
+				}
+				newPath = newPath.replace(tag, data[tagName]);
+			});
+		});
+
+		// Look for optionals. There should be a max of one as they extend a query
+		// string.
+		reSearch(/\[(.*?)\]/g, (optionalTags: RegExpMatchArray) => {
+			if (optionalTags.length > 1) {
+				throw new Error(`Front endpoint ${path} is incorrectly defined`);
+			}
+
+			// Add each of these tags to an array used for the query string.
+			const trimmedTags = optionalTags[0];
+			const tags = trimmedTags.substring(1, trimmedTags.length - 1).split(':');
+			const queryTags: { [key: string]: string } = {};
+
+			// Ensure we remove the optional signature.
+			newPath = newPath.replace(trimmedTags, '');
+			_.each(tags, (tag) => {
+				if (data[tag]) {
+					queryTags[tag] = data[tag];
+				}
+			});
+			newPath = `${newPath}?${querystring.stringify(queryTags)}`;
+		});
+
+		return newPath;
+	}
+}
+
+export type Callback<T> = (err: Error | null, response: T | null) => void;
+
+export interface Attachment {
+	filename: string;
+	url: string;
+	contentType: string;
+	size: number;
+	metadata: any;
+}
+
+export interface Author {
+	_links: Links;
+	id: string;
+	email: string;
+	username: string;
+	first_name: string;
+	last_name: string;
+	is_admin: boolean;
+	is_available: boolean;
```

Comment:
Because I'd have to convert to/from every single response we receive from Front. This can potentially lead to foulups when their interfaces change or we miss something. I originally did this, and realised it was dangerous when I found a name issue in their interface (which they've now fixed after reporting it).